File: test_chat.mjs
#!/usr/bin/env node
/**
 * test_chat.mjs — Node.js test client for web-model-bridge OpenAI-compatible API
 *
 * Usage:
 *   node test_chat.mjs
 *   node test_chat.mjs --url http://127.0.0.1:3457/v1
 *   node test_chat.mjs --model gemini --prompt "What is speed of light?"
 *   node test_chat.mjs --model chatgpt --prompt "Hello from ChatGPT"
 *   node test_chat.mjs --list-models --mode providers
 *   node test_chat.mjs --list-models --mode models
 *   node test_chat.mjs --discover
 *   node test_chat.mjs --file ./image.png --prompt "What is in this file?"
 */

import { readFileSync, existsSync } from 'node:fs';
import { basename } from 'node:path';

const DEFAULT_URL = 'http://127.0.0.1:3457/v1';
const FALLBACK_URL = 'http://127.0.0.1:3456/v1';
const DEFAULT_MODEL = 'gemini';
const DEFAULT_KEY = 'not-needed';

// Parse command line arguments
const args = process.argv.slice(2);
function getArg(flag, defaultValue) {
  const idx = args.indexOf(flag);
  if (idx !== -1 && idx + 1 < args.length) return args[idx + 1];
  return defaultValue;
}
const hasFlag = (flag) => args.includes(flag);

const preferredUrl = getArg('--url', DEFAULT_URL).replace(/\/+$/, '');
const model = getArg('--model', DEFAULT_MODEL);
const apiKey = getArg('--key', DEFAULT_KEY);
const prompt = getArg('--prompt', 'Hello! Tell me 2 interesting space facts.');
const mode = getArg('--mode', '');
const filePath = getArg('--file', '');
const isListModels = hasFlag('--list-models');
const isDiscover = hasFlag('--discover');
const isNonStream = hasFlag('--no-stream');

async function probeBaseUrl(targetUrl) {
  const candidates = [targetUrl];
  if (targetUrl === DEFAULT_URL) candidates.push(FALLBACK_URL);
  else if (targetUrl === FALLBACK_URL) candidates.push(DEFAULT_URL);

  for (const candidate of candidates) {
    try {
      const res = await fetch(`${candidate}/models`, { signal: AbortSignal.timeout(2000) });
      if (res.ok) return candidate;
    } catch {
      // try next
    }
  }
  return targetUrl;
}

async function listModels(baseUrl) {
  const url = mode ? `${baseUrl}/models?mode=${mode}` : `${baseUrl}/models`;
  console.log(`\n[*] Querying models from: ${url}`);
  try {
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${res.statusText}`);
    const data = await res.json();
    console.log(`[+] Retrieved ${data.data?.length || 0} model(s):`);
    for (const m of data.data || []) {
      console.log(`    * ${m.id} ${m.name ? `(${m.name})` : ''}`);
    }
  } catch (err) {
    console.error(`[-] Failed to fetch models: ${err.message}`);
  }
}

async function discoverModels(baseUrl) {
  console.log(`\n[*] Triggering model discovery at: ${baseUrl}/models/discover`);
  try {
    const res = await fetch(`${baseUrl}/models/discover`, {
      method: 'POST',
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${res.statusText}`);
    const data = await res.json();
    console.log(`[+] Discovered ${data.count || 0} active models:`);
    for (const m of data.models || []) {
      console.log(`    * ${m.id} - ${m.name}`);
    }
  } catch (err) {
    console.error(`[-] Discovery failed: ${err.message}`);
  }
}

async function chat(baseUrl) {
  const endpoint = `${baseUrl}/chat/completions`;
  const stream = !isNonStream;

  console.log(`\n[*] Target Endpoint: ${endpoint}`);
  console.log(`[*] Model: ${model}`);
  console.log(`[*] Stream: ${stream}`);
  console.log(`[*] Prompt: "${prompt.slice(0, 100)}${prompt.length > 100 ? '...' : ''}"`);
  
  const files = [];
  if (filePath) {
    if (existsSync(filePath)) {
      const buffer = readFileSync(filePath);
      const ext = filePath.split('.').pop()?.toLowerCase() || '';
      const mime = ext === 'png' ? 'image/png'
        : ext === 'jpg' || ext === 'jpeg' ? 'image/jpeg'
        : ext === 'pdf' ? 'application/pdf'
        : ext === 'mp4' ? 'video/mp4'
        : ext === 'mp3' ? 'audio/mp3'
        : 'text/plain';

      files.push({
        name: basename(filePath),
        type: mime,
        data: `data:${mime};base64,${buffer.toString('base64')}`,
      });
      console.log(`[*] Attached file: ${basename(filePath)} (${buffer.length} bytes)`);
    } else {
      console.warn(`[-] File not found: ${filePath}`);
    }
  }

  console.log('-'.repeat(50));
  process.stdout.write('Assistant: ');

  try {
    const res = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model,
        messages: [{ role: 'user', content: prompt }],
        files,
        stream,
      }),
    });

    if (!res.ok) {
      const text = await res.text();
      console.log(`\n[-] HTTP ${res.status}: ${text}`);
      return;
    }

    if (!stream) {
      const data = await res.json();
      const content = data.choices?.[0]?.message?.content || '(No content)';
      console.log(content);
      if (data.files && data.files.length > 0) {
        console.log(`[+] Generated files:`, data.files);
      }
      console.log('-'.repeat(50));
      console.log(`[+] Response ID: ${data.id}`);
      return;
    }

    // Stream SSE chunks
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let accumulated = '';
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.startsWith('data:')) continue;
        const dataStr = trimmed.slice(5).trim();
        if (dataStr === '[DONE]') continue;

        try {
          const chunk = JSON.parse(dataStr);
          if (chunk.files) {
            console.log(`\n[+] Received files:`, chunk.files);
          }
          if (chunk.error_scene) {
            console.log(`\n[!] Error Scene captured:`, chunk.error_scene.url);
          }
          const delta = chunk.choices?.[0]?.delta?.content || '';
          if (delta) {
            process.stdout.write(delta);
            accumulated += delta;
          }
        } catch {
          // ignore
        }
      }
    }

    console.log('\n' + '-'.repeat(50));
    console.log(`[+] Stream finished (${accumulated.length} chars).`);
  } catch (err) {
    console.error(`\n[-] Connection / Request failed: ${err.message}`);
    console.log(`    Make sure web-model-bridge is running on ${baseUrl}`);
  }
}

async function main() {
  const baseUrl = await probeBaseUrl(preferredUrl);
  if (baseUrl !== preferredUrl) {
    console.log(`[*] Auto-switched port: ${preferredUrl} -> ${baseUrl}`);
  }

  console.log('='.repeat(50));
  console.log(`[WEB-MODEL-BRIDGE] Hackathon Test Client`);
  console.log(`API Base: ${baseUrl}`);
  console.log('='.repeat(50));

  if (isDiscover) {
    await discoverModels(baseUrl);
  } else if (isListModels) {
    await listModels(baseUrl);
  } else {
    await chat(baseUrl);
  }
}

main();

File: test_chat.py
#!/usr/bin/env python3
"""
test_chat.py — Test client for web-model-bridge OpenAI-compatible API
Sends chat completion requests to http://127.0.0.1:3457/v1 (or 3456)
Usage:
    python test_chat.py
    python test_chat.py --model gemini-web/gemini-3-flash
    python test_chat.py --model gemini-web/gemini-2.5-pro --prompt "Explain black holes"
    python test_chat.py --list-models
    python test_chat.py --no-stream
"""

import sys
import json
import argparse

# Force UTF-8 on Windows console if supported
if hasattr(sys.stdout, 'reconfigure'):
    try:
        sys.stdout.reconfigure(encoding='utf-8')
    except Exception:
        pass
try:
    import requests
except ImportError:
    requests = None
    import urllib.request
    import urllib.error

DEFAULT_URL = "http://127.0.0.1:3457/v1"
FALLBACK_URL = "http://127.0.0.1:3456/v1"
DEFAULT_MODEL = "gemini-web/gemini-3-flash"
DEFAULT_KEY = "not-needed"


def probe_base_url(preferred_url):
    """Check if preferred URL is listening, otherwise try fallback."""
    urls_to_try = [preferred_url]
    if preferred_url == DEFAULT_URL:
        urls_to_try.append(FALLBACK_URL)
    elif preferred_url == FALLBACK_URL:
        urls_to_try.append(DEFAULT_URL)

    for base in urls_to_try:
        clean = base.rstrip('/')
        test_url = f"{clean}/models"
        try:
            if requests:
                res = requests.get(test_url, timeout=2)
                if res.status_code == 200:
                    return clean
            else:
                req = urllib.request.Request(test_url)
                with urllib.request.urlopen(req, timeout=2) as r:
                    if r.status == 200:
                        return clean
        except Exception:
            continue
    return preferred_url.rstrip('/')


def list_models(base_url, api_key):
    url = f"{base_url}/models"
    headers = {"Authorization": f"Bearer {api_key}"}
    print(f"\n[*] Querying models from: {url}")
    try:
        if requests:
            r = requests.get(url, headers=headers, timeout=5)
            r.raise_for_status()
            data = r.json()
        else:
            req = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(req, timeout=5) as resp:
                data = json.loads(resp.read().decode('utf-8'))

        models = data.get("data", [])
        print(f"[+] Successfully retrieved {len(models)} model(s):")
        for m in models:
            m_id = m.get("id", "unknown")
            print(f"    * {m_id}")
        return models
    except Exception as e:
        print(f"[-] Failed to list models: {e}")
        return []


def chat_stream(base_url, model, messages, api_key):
    url = f"{base_url}/chat/completions"
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}"
    }
    payload = {
        "model": model,
        "messages": messages,
        "stream": True
    }

    print(f"\n[*] Sending streaming request to: {url}")
    print(f"[*] Model: {model}")
    print(f"[*] Prompt: \"{messages[-1]['content']}\"")
    print("-" * 50)
    print("Assistant: ", end="", flush=True)

    if not requests:
        print("\n(requests library not found, falling back to non-streaming)")
        return chat_non_stream(base_url, model, messages, api_key)

    try:
        with requests.post(url, headers=headers, json=payload, stream=True, timeout=60) as resp:
            if resp.status_code != 200:
                print(f"\n[✗] HTTP {resp.status_code}: {resp.text}")
                return

            full_reply = ""
            for line in resp.iter_lines(decode_unicode=True):
                if not line:
                    continue
                if line.startswith("data:"):
                    data_str = line[5:].strip()
                    if data_str == "[DONE]":
                        break
                    try:
                        chunk = json.loads(data_str)
                        delta = chunk.get("choices", [{}])[0].get("delta", {}).get("content", "")
                        if delta:
                            print(delta, end="", flush=True)
                            full_reply += delta
                    except json.JSONDecodeError:
                        continue

            print("\n" + "-" * 50)
            print(f"[+] Stream completed ({len(full_reply)} characters received).")

    except requests.exceptions.ConnectionError:
        print(f"\n[-] Could not connect to {url}.")
        print("    Please ensure web-model-bridge is running (run 'npm start').")
    except Exception as e:
        print(f"\n[-] Error during stream: {e}")


def chat_non_stream(base_url, model, messages, api_key):
    url = f"{base_url}/chat/completions"
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}"
    }
    payload = {
        "model": model,
        "messages": messages,
        "stream": False
    }

    print(f"\n[*] Sending non-streaming request to: {url}")
    print(f"[*] Model: {model}")
    print(f"[*] Prompt: \"{messages[-1]['content']}\"")
    print("-" * 50)

    try:
        if requests:
            resp = requests.post(url, headers=headers, json=payload, timeout=60)
            if resp.status_code != 200:
                print(f"[-] HTTP {resp.status_code}: {resp.text}")
                return
            data = resp.json()
        else:
            req_data = json.dumps(payload).encode('utf-8')
            req = urllib.request.Request(url, data=req_data, headers=headers, method='POST')
            with urllib.request.urlopen(req, timeout=60) as r:
                data = json.loads(r.read().decode('utf-8'))

        content = data.get("choices", [{}])[0].get("message", {}).get("content", "")
        print(f"Assistant: {content}")
        print("-" * 50)
        print(f"[+] Non-stream completed (ID: {data.get('id', 'N/A')}).")
    except Exception as e:
        print(f"[-] Error: {e}")


def main():
    parser = argparse.ArgumentParser(description="Test chat client for web-model-bridge")
    parser.add_argument("--url", default=DEFAULT_URL, help=f"API Base URL (default: {DEFAULT_URL})")
    parser.add_argument("--model", default=DEFAULT_MODEL, help=f"Model ID (default: {DEFAULT_MODEL})")
    parser.add_argument("--key", default=DEFAULT_KEY, help=f"API Key / Bearer token (default: {DEFAULT_KEY})")
    parser.add_argument("--prompt", default="Hello! Please introduce yourself in 2 concise sentences.", help="Prompt text")
    parser.add_argument("--system", default="", help="Optional system prompt")
    parser.add_argument("--no-stream", action="store_true", help="Disable streaming response")
    parser.add_argument("--list-models", action="store_true", help="List available models and exit")
    parser.add_argument("--no-auto-probe", action="store_true", help="Don't auto-probe between port 3457 and 3456")

    args = parser.parse_args()

    base_url = args.url
    if not args.no_auto_probe:
        probed = probe_base_url(args.url)
        if probed != args.url.rstrip('/'):
            print(f"[*] Port auto-detected: switching from {args.url} to {probed}")
        base_url = probed
    else:
        base_url = base_url.rstrip('/')

    print(f"==================================================")
    print(f"[WEB-MODEL-BRIDGE] Test Client")
    print(f"Target API: {base_url}")
    print(f"==================================================")

    if args.list_models:
        list_models(base_url, args.key)
        return

    messages = []
    if args.system:
        messages.append({"role": "system", "content": args.system})
    messages.append({"role": "user", "content": args.prompt})

    if args.no_stream:
        chat_non_stream(base_url, args.model, messages, args.key)
    else:
        chat_stream(base_url, args.model, messages, args.key)


if __name__ == "__main__":
    main()
